fix(postgres/mssql/saphana): include primary key in upsert

CHOIJEWON opened the pull request:
### Description of change

**Fixes #11556**

This PR addresses multiple related issues with upsert operations when using identity/auto-increment columns as conflict
paths across PostgreSQL, MSSQL, and SAP HANA.

## 🐛 Problems

### 1. PostgreSQL: Duplicate Rows on Primary Key Upsert

When using `upsert()` with a primary key as the conflict path in PostgreSQL, the method creates duplicate rows instead
of updating existing ones.

**Reproduction:**

```typescript
// First upsert with id=1
await postRepository.upsert(
    { id: 1, title: "First" },
    { conflictPaths: ['id'] }
);

// Second upsert with same id=1
await postRepository.upsert(
    { id: 1, title: "Updated" },
    { conflictPaths: ['id'] }
);

// Expected: 1 row with updated title
// Actual: 2 rows created (conflict detection failed)
```

### 2. SAP HANA/MSSQL: Identity Columns as Conflict Paths Fail

MERGE INTO queries fail when using identity columns as conflict paths:

- **SAP HANA**: `cannot insert generated identity column field id`
- **MSSQL**: `Cannot insert explicit value for identity column when IDENTITY_INSERT is OFF`

### 3. MSSQL: IDENTITY_INSERT Not Applied to MERGE Queries

When explicitly providing values for identity columns in MERGE queries:

- Error: `DEFAULT not allowed for IDENTITY columns in MERGE`
- `SET IDENTITY_INSERT` was only applied to INSERT queries, not MERGE

## 💡 Root Causes

### Issue 1: PostgreSQL Auto-Increment Exclusion

In `getInsertedColumns()`, auto-increment primary keys were filtered out for PostgreSQL, even when users explicitly
provided values. This caused the ON CONFLICT clause to fail.

### Issue 2: MERGE INTO Source Missing Conflict Columns

In `createMergeIntoSourceExpression()`, identity columns were excluded from the USING clause, preventing the ON clause
from matching correctly.

Additionally, `createMergeIntoInsertValuesExpression()` tried to insert explicit values into identity columns, which is
prohibited in SAP HANA and MSSQL.

### Issue 3: Missing IDENTITY_INSERT for MERGE

The `createMergeExpression()` method lacked the IDENTITY_INSERT wrapper that was already present in
`createInsertExpression()`.

## ✅ Solutions

### Fix 1: Include Primary Keys in PostgreSQL Upserts

Applied the same `isOverridingAutoIncrementBehavior()` logic used for MSSQL to PostgreSQL:

```typescript
// src/query-builder/InsertQueryBuilder.ts:746-754
if (
    column.isGenerated &&
    column.generationStrategy === "increment" &&
    // ... other conditions
    !(
        (this.connection.driver.options.type === "mssql" ||
            DriverUtils.isPostgresFamily(this.connection.driver)) &&
        this.isOverridingAutoIncrementBehavior(column)
    )
)
```

### Fix 2: Support Identity Columns in MERGE Conflict Paths

**2a. Include conflict columns in USING clause:**

```typescript
// createMergeIntoSourceExpression()
// Add conflict path columns even if auto-generated
// (needed for ON clause matching)
if (this.expressionMap.onUpdate?.conflict) {
    const conflictColumns = // ... find conflict columns
        columns = [...columns, ...additionalColumns]
}
```

**2b. Use DEFAULT for identity columns in INSERT values:**

```typescript
// createMergeIntoInsertValuesExpression()
if (
    (column.isGenerated &&
        column.generationStrategy !== "uuid" &&
        !this.isOverridingAutoIncrementBehavior(column))
) {
    expression += `DEFAULT`
}
```

### Fix 3: Add IDENTITY_INSERT to MERGE Queries

Mirrors the existing IDENTITY_INSERT handling from `createInsertExpression()`:

```typescript
// createMergeExpression()
if (
    this.connection.driver.options.type === "mssql" &&
    this.expressionMap.mainAlias!.hasMetadata &&
    columns.some((column) =>
        this.isOverridingAutoIncrementBehavior(column),
    )
) {
    query = `SET IDENTITY_INSERT ${tableName} ON; ${query}; SET IDENTITY_INSERT ${tableName} OFF`
}
```

## 🧪 Testing

**New test case added:**

```typescript
it("should upsert with primary key as conflict path", () => {
    // Tests that upsert with id as conflict path updates
    // instead of creating duplicates
})
```

**Test results:**

```bash
# New test
npm run test:fast -- --grep "should upsert with primary key as conflict path"
✔ should upsert with primary key as conflict path

# All upsert tests
npm run test:fast -- --grep "upsert"
✔ 13 passing

# All repository basic methods tests
npm run test:fast -- --grep "repository > basic methods"
✔ 27 passing
```

---

## 🔍 Impact Analysis

### Affected Drivers

- ✅ **PostgreSQL/CockroachDB**: Can now use primary keys as conflict paths in upserts
- ✅ **MSSQL**: MERGE queries with identity columns now work correctly
- ✅ **SAP HANA**: MERGE queries with identity columns now work correctly
- ⚪ **No impact**: MySQL, MariaDB, SQLite, Oracle (no behavioral changes)

### Breaking Changes

**None.** These are bug fixes that enable previously broken functionality.

##  Pull-Request Checklist

- [x] Code is up-to-date with the `master` branch
- [x] This pull request links relevant issues as `Fixes #11556`
- [x] There are new or updated tests validating the change
- [x] All existing tests pass
- [ ] Documentation has been updated to reflect this change - N/A (bug fixes, no API changes)

CHOIJEWON committed:
fix(postgres): include primary key in upsert when explicitly provided (#11556)

When using upsert() with a primary key as the conflict path, PostgreSQL now
correctly updates existing rows instead of creating duplicates.

This aligns PostgreSQL behavior with MSSQL by including auto-increment
primary keys in INSERT statements when users explicitly provide ID values.

gioboa reviewed (commented):
Some tests failed, I merged develop into this branch to align it and trigger the tests

gioboa reviewed (dismissed):
Few tests are failing, can you verify them please? @CHOIJEWON

gioboa reviewed (commented):
I see, option A is the best one I guess.
Btw we can't merge a PR with failed tests, the button is disabled 🙈
Thanks for your great help 🙏🚀

CHOIJEWON committed:
fix(merge): support identity columns as conflict paths in MERGE INTO

  Both SAP HANA and MSSQL do not allow inserting explicit values into
  IDENTITY columns. This commit fixes MERGE INTO upsert when using
  primary keys or other identity columns as conflict paths.

  The fix ensures identity columns are:
  1. Included in USING clause (needed for ON clause conflict detection)
  2. Set to DEFAULT in INSERT clause (instead of explicit values)

  Changes:
  - createMergeIntoSourceExpression: Add conflict path columns to USING
    clause even when auto-generated (for ON clause matching)
  - createMergeIntoInsertValuesExpression: Check isOverridingAutoIncrementBehavior
    before inserting values, use DEFAULT for identity columns

  Fixes errors:
  - SAP HANA: 'cannot insert generated identity column field id'
  - MSSQL: 'Cannot insert explicit value for identity column when IDENTITY_INSERT is OF'

CHOIJEWON committed:
fix(mssql): enable IDENTITY_INSERT for MERGE queries with identity columns

  Fixes 'DEFAULT not allowed for IDENTITY columns in MERGE' error when
  explicitly providing values for auto-increment primary keys in MERGE
  queries. This mirrors the existing IDENTITY_INSERT handling in
  createInsertExpression() for consistency.

  The SET IDENTITY_INSERT wrapper is only applied when:
  - Using MSSQL driver
  - Entity has metadata
  - User explicitly provides values for identity columns

gioboa reviewed (changes requested):
is this PR still valid? can you solve the conflicts please?

CHOIJEWON commented:
@gioboa 

I’ve resolved the merge conflicts, but I’m encountering a status 522 error during the Coveralls parallel build step.
Could you please check if this is something on the CI/Coveralls side?

gioboa reviewed (commented):
Manually triggered the pipeline 🤞

CHOIJEWON commented:
@gioboa 

The tests are passing successfully, but the Coveralls Parallel step continues to fail 😭. Is it possible that this is related to my code, or might there be another cause? I’d appreciate any guidance or suggestions.

gioboa reviewed (commented):
Coverall infrastructure is down since yesterday so it's not your fault, we need to wait until the problem is fixed on their side 🥲

Commit:
Merge remote-tracking branch 'origin/master' into work-11820

# Conflicts:
#	src/query-builder/InsertQueryBuilder.ts
